rse-py-scripting.Rmd

```diff
@@ -0,0 +1,143 @@
+# Command line programs {#rse-py-scripting}
+
+## Questions {#rse-py-scripting-questions}
+
+```{r, child="questions/rse-py-scripting.md"}
+```
+
+
+## Introduction {#rse-py-scripting-intro}
+
+The Jupyter Notebook, PyCharm and other interactive tools
+are great for prototyping code and exploring data,
+but in many cases you will ultimately want to take the code you've developed
+using those tools and apply it to thousands of data files, 
+run it with many different parameters,
+and/or combine it with other programs in a data analysis pipeline. 
+In order to do these things,
+it is often necessary/preferable to turn the code you've developed into a program 
+that works like any other Unix command-line tool.
+
+In this chapter we take the code we developed earlier (TODO: Link)
+and turn it into a Python script (i.e. a command line program written in Python).
+
+
+## A template script {#rse-py-scripting-template}
+
+In general, the first thing that gets added to any Python script is the following:
+
+```python
+if __name__ == '__main__':
+    main()
+```
+
+The reason we need these two lines of code
+is that running a Python script in bash is very similar to importing that file in Python. 
```

Comment:
I remember struggling with this when I started learning Python... I think we could make these two paragraphs slightly clearer, what do you think of something along these lines?

```suggestion
The reason we need these two lines of code
is to differentiate between running a Python file as a standalone program
and importing it as a module. 
When you import a Python file as a module,
the `__name__` variable is automatically set to the name of the file
(e.g. when importing script.py, `__name__` is `script`).
When running a Python file as a standalone program,
`__name__` is always set to `__main__`.
This means that we can separate the two cases above by checking the value of `__name__`.
When the Python file is running as a standalone program,
you want to execute its main functionality.
You conventionally define what this is inside a function named `main()`,
but you could call this function whatever you prefer.
```